### PR TITLE
allow eth.filter to return appropriate sync/async filter types

### PR DESCRIPTION
**Pull request by pacrob**
Calls to AsyncEth.filter were returning sync filter types

### How was it fixed?

Updated associated plumbing, added testing

#### Cute Animal Picture

image

### CODE REVIEW

**Review by kclowes — COMMENTED** [review]
`Module` has a property called `is_async` already. I _think_ we could use that to differentiate instead of adding another `is_async` property on `Method`. 

Also, I haven't tested this, but it looks like you could also combine the `filter_wrapper` with `async_filter_wrapper` if you wanted to. That's more of a stylistic thing though :) Otherwise, this looks good!

File: web3/_utils/method_formatters.py [hunk 1]
```diff
@@ -754,42 +754,87 @@ def raise_transaction_not_found_with_index(
...
+) -> Union[
+    AsyncBlockFilter,
+    AsyncTransactionFilter,
+    AsyncLogFilter,
+    BlockFilter,
+    TransactionFilter,
+    LogFilter,
+]:
+    
+    if type(module).__name__ == "Eth":
```
Comment (kclowes) [inline web3/_utils/method_formatters.py:789]: Sorry if this is just a wip and you don't want feedback, but could you check for `module.is_async` here?

Comment (pacrob) [reply web3/_utils/method_formatters.py:789]: No problem - feedback always welcome, especially if I've asked for review. Will try it.

Comment (pacrob) [reply web3/_utils/method_formatters.py]: Ok, updated, ready for re-review.

**Review by kclowes — APPROVED** [review]
Nice! LGTM!

**Pull request merged**